light-client/src/types.rs

```diff
@@ -74,46 +66,115 @@ impl Status {
     }
 }
 
+pub trait Commit {
+    fn block_hash(&self) -> Hash;
+}
+
+pub trait LightBlock: Send + Sync + Clone + Debug + Serialize + DeserializeOwned + 'static {
+    type Header;
+    type Commit: Commit;
+    type SignedHeader: Clone;
+    type ValidatorSet;
+    type Evidence;
+
+    fn height(&self) -> Height;
+    fn signed_header(&self) -> &Self::SignedHeader;
+    fn header(&self) -> &Self::Header;
+    fn commit(&self) -> &Self::Commit;
+    fn validators(&self) -> &Self::ValidatorSet;
+    fn next_validators(&self) -> &Self::ValidatorSet;
+    fn header_time(&self) -> Time;
+    fn validators_hash(&self) -> Hash;
+    fn next_validators_hash(&self) -> Hash;
+    fn provider(&self) -> PeerId;
+}
+
+impl Commit for TMCommit {
+    fn block_hash(&self) -> Hash {
+        self.block_id.hash
+    }
+}
+
+impl LightBlock for TMLightBlock {
```

Comment:
Definitely looks like a good first step. I think what we are looking for is to simplify the object graph to the point where it's easy to test and be general over changes to the underlying data structure. I think what might be good here is actually look not just at the source of the dependency graph (TMLightBlock) but the dependencies. There are so many uses of the LightBlock that are related to internal validation, for instance insuring the validator set hashes actually match the validator set. It these cases it seems like we could simplify the dependency graph and uses by having a separate type, ValidLightBlock which provides these guarantees at construction. 

I ask raise this here as it isn't entirely clear to me how much of the underlying data structures we need to expose here. Do we need the entire `signed_header` or just properties of it? I know this goes against my argument that motivated the `operations` I'm just wondering if it's not worthwhile to revisit.

Reply:
>  I think what might be good here is to look not just at the source of the dependency graph (TMLightBlock) but the uses. 

Yeah that's actually what I did, I only added methods to the `LightBlock` trait when they were required at use site. But that doesn't mean that this interface is as minimal as it could be, as the code at use site was written with the concrete `TMLightBlock` in mind.

> There are so many uses the LightBlock that are related to internal validation, for instance ensuring the validator set hashes actually match the validator set. In these cases it may be simpler have a separate type, ValidLightBlock which provides these guarantees at construction.

Interesting idea, I definitely want to explore encoding more invariants in the types themselves.

> I raise this here as it isn't entirely clear to me how much of the underlying data structures we need to expose. Do we need the entire signed_header or just properties of it?

Good question! For `SignedHeader`, I don't believe it's needed anymore, as it was only used to implement header hashing, voting power calculation and commit validation via the `Commit` trait, which has been split up in separate traits.

Reply:
Actually, `SignedHeader` is currently also used in the [`ConflictingHeadersEvidence` struct](https://github.com/informalsystems/tendermint-rs/blob/2112202e79a77ded10114733138675cceb60cc60/tendermint/src/evidence.rs#L39-L44), but could easily be removed by introducing a new private struct to the evidence module with the same definition.